File: app/src/main/java/ch/threema/app/globalsearch/GlobalSearchActivity.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2019-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.globalsearch;

import static com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_DRAGGING;
import static com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_EXPANDED;
import static com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_HIDDEN;
import static com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_SETTLING;
import static ch.threema.app.services.MessageServiceImpl.FILTER_CHATS;
import static ch.threema.app.services.MessageServiceImpl.FILTER_GROUPS;
import static ch.threema.app.services.MessageServiceImpl.FILTER_INCLUDE_ARCHIVED;

import android.annotation.SuppressLint;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.UiThread;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.chip.Chip;
import com.google.android.material.progressindicator.CircularProgressIndicator;
import com.google.android.material.shape.MaterialShapeDrawable;
import com.google.android.material.shape.ShapeAppearanceModel;

import org.slf4j.Logger;

import ch.threema.app.R;
import ch.threema.app.activities.ThreemaActivity;
import ch.threema.app.activities.ThreemaToolbarActivity;
import ch.threema.app.collections.Functional;
import ch.threema.app.collections.IPredicateNonNull;
import ch.threema.app.services.ContactService;
import ch.threema.app.services.DeadlineListService;
import ch.threema.app.services.GroupService;
import ch.threema.app.services.MessageService;
import ch.threema.app.ui.ThreemaSearchView;
import ch.threema.app.utils.EditTextUtil;
import ch.threema.app.utils.IntentDataUtil;
import ch.threema.base.utils.LoggingUtil;
import ch.threema.storage.models.AbstractMessageModel;
import ch.threema.storage.models.GroupMessageModel;

public class GlobalSearchActivity extends ThreemaToolbarActivity implements ThreemaSearchView.OnQueryTextListener {
	private static final Logger logger = LoggingUtil.getThreemaLogger("GlobalSearchActivity");
	private static final int QUERY_MIN_LENGTH = 2;
	private static final long GLOBAL_SEARCH_QUERY_TIMEOUT_MS = 500;

	private GlobalSearchAdapter chatsAdapter;
	private GlobalSearchViewModel chatsViewModel;
	private TextView emptyTextView;
	private ThreemaSearchView searchView;
	private CircularProgressIndicator progressBar;
	private DeadlineListService hiddenChatsListService;
	private ContactService contactService;
	private GroupService groupService;

	private @MessageService.MessageFilterFlags int filterFlags = FILTER_CHATS | FILTER_GROUPS | FILTER_INCLUDE_ARCHIVED;
	private String queryText;
	private final Handler queryHandler = new Handler();
	private final Runnable queryTask = new Runnable() {
		@Override
		public void run() {
			chatsViewModel.onQueryChanged(queryText, filterFlags, false, false);
			chatsAdapter.onQueryChanged(queryText);
		}
	};

	@Override
	public boolean onQueryTextSubmit(String query) {
		// Do something
		return true;
	}

	@Override
	public boolean onQueryTextChange(String newText) {
		queryText = newText;

		if (chatsViewModel != null && chatsAdapter != null) {

			queryHandler.removeCallbacksAndMessages(null);
			if (queryText != null && queryText.length() >= QUERY_MIN_LENGTH) {
				queryHandler.postDelayed(queryTask, GLOBAL_SEARCH_QUERY_TIMEOUT_MS);
			} else {
				chatsViewModel.onQueryChanged(null, filterFlags, false, false);
				chatsAdapter.onQueryChanged(null);
			}
		}

		return true;
	}

	public int getLayoutResource() {
		return R.layout.activity_global_search;
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		try {
			contactService = serviceManager.getContactService();
			groupService = serviceManager.getGroupService();
			hiddenChatsListService = serviceManager.getHiddenChatsListService();
		} catch (Exception e) {
			logger.error("Exception", e);
			finish();
		}
	}

	@SuppressLint("ClickableViewAccessibility")
	@Override
	protected boolean initActivity(Bundle savedInstanceState) {
		if (!super.initActivity(savedInstanceState)) {
			return false;
		};

		ConstraintLayout bottomSheetLayout = findViewById(R.id.bottom_sheet);
		final float cornerSize = getResources().getDimensionPixelSize(R.dimen.bottomsheet_corner_size);

		final BottomSheetBehavior<ConstraintLayout> bottomSheetBehavior = BottomSheetBehavior.from(bottomSheetLayout);
		bottomSheetBehavior.addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
			@Override
			public void onStateChanged(@NonNull View bottomSheet, int newState) {
				Drawable background = bottomSheetLayout.getBackground();

				switch (newState) {
					case STATE_HIDDEN:
						finish();
						break;
					case STATE_EXPANDED:
						findViewById(R.id.drag_handle).setVisibility(View.INVISIBLE);
						if (background instanceof MaterialShapeDrawable) {
							MaterialShapeDrawable materialShapeDrawable = (MaterialShapeDrawable) background;
							getWindow().setStatusBarColor(materialShapeDrawable.getResolvedTintColor());
							ShapeAppearanceModel shapeAppearanceModel = materialShapeDrawable.getShapeAppearanceModel().toBuilder()
								.setAllCornerSizes(0)
								.build();
							materialShapeDrawable.setShapeAppearanceModel(shapeAppearanceModel);
						} else {
							getWindow().setStatusBarColor(getResources().getColor(R.color.attach_status_bar_color_expanded));
						}
						break;
					case STATE_SETTLING:
						findViewById(R.id.drag_handle).setVisibility(View.VISIBLE);
						break;
					case STATE_DRAGGING:
						getWindow().setStatusBarColor(getResources().getColor(R.color.attach_status_bar_color_collapsed));
						if (background instanceof MaterialShapeDrawable) {
							MaterialShapeDrawable materialShapeDrawable = (MaterialShapeDrawable) background;
							materialShapeDrawable.setShapeAppearanceModel(
								materialShapeDrawable.getShapeAppearanceModel().toBuilder()
									.setTopLeftCornerSize(cornerSize)
									.setTopRightCornerSize(cornerSize)
									.build()
							);
						}
					default:
						break;
				}
			}

			@Override
			public void onSlide(@NonNull View bottomSheet, float slideOffset) {
				// we don't care about sliding events
			}
		});

		findViewById(R.id.parent_layout).setOnClickListener(v -> bottomSheetBehavior.setState(STATE_HIDDEN));

		getWindow().setStatusBarColor(getResources().getColor(R.color.attach_status_bar_color_collapsed));

		searchView = findViewById(R.id.search);
		searchView.setOnQueryTextListener(this);

		emptyTextView = findViewById(R.id.empty_text);
		progressBar = findViewById(R.id.progress);

		chatsAdapter = new GlobalSearchAdapter(
			this,
			Glide.with(this),
			R.layout.item_global_search,
			17
		);
		chatsAdapter.setOnClickItemListener((messageModel, view, position) -> showMessage(messageModel, view));

		setupChip(R.id.chats, FILTER_CHATS, true);
		setupChip(R.id.groups, FILTER_GROUPS, true);
		setupChip(R.id.archived, FILTER_INCLUDE_ARCHIVED, true);

		RecyclerView chatsRecyclerView = this.findViewById(R.id.recycler_chats);
		chatsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
		chatsRecyclerView.setItemAnimator(new DefaultItemAnimator());
		chatsRecyclerView.setAdapter(chatsAdapter);

		chatsViewModel = new ViewModelProvider(this).get(GlobalSearchViewModel.class);
		chatsViewModel.getMessageModels().observe(this, messageModels -> {
			if (messageModels.size() > 0) {
				messageModels = Functional.filter(messageModels, (IPredicateNonNull<AbstractMessageModel>) messageModel -> {
					if (messageModel instanceof GroupMessageModel) {
						if (((GroupMessageModel) messageModel).getGroupId() > 0) {
							return !hiddenChatsListService.has(groupService.getUniqueIdString(((GroupMessageModel) messageModel).getGroupId()));
						}
					} else {
						if (messageModel.getIdentity() != null) {
							return !hiddenChatsListService.has(contactService.getUniqueIdString(messageModel.getIdentity()));
						}
					}
					return true;
				});
			}

			if (messageModels.size() == 0) {
				if (queryText != null && queryText.length() >= QUERY_MIN_LENGTH) {
					emptyTextView.setText(R.string.search_no_matches);
				} else {
					emptyTextView.setText(R.string.global_search_empty_view_text);
				}
				emptyTextView.setVisibility(View.VISIBLE);
			} else {
				emptyTextView.setVisibility(View.GONE);
			}
			chatsAdapter.setMessageModels(messageModels);
		});

		chatsViewModel.getIsLoading().observe(this, isLoading -> {
			if (isLoading != null) {
				showProgressBar(isLoading);
			}
		});
		return true;
	}

	private void setupChip(@IdRes int id, int flag, boolean checked) {
		// https://github.com/material-components/material-components-android/issues/1419
		Chip chip = findViewById(id);
		chip.setChecked(checked);
		chip.setOnCheckedChangeListener((buttonView, isChecked) -> {
			if (isChecked) {
				filterFlags |= flag;
			} else {
				filterFlags  &= ~flag;
			}
			chatsViewModel.onQueryChanged(queryText, filterFlags, false, false);
		});
	}

	private void showMessage(AbstractMessageModel messageModel, View view) {
		if (messageModel == null) {
			return;
		}

		if (searchView != null) {
			EditTextUtil.hideSoftKeyboard(searchView);
		}

		startActivityForResult(IntentDataUtil.getJumpToMessageIntent(this, messageModel), ThreemaActivity.ACTIVITY_ID_COMPOSE_MESSAGE);

		finish();
	}

	@UiThread
	synchronized private void showProgressBar(boolean show) {
		if (show) {
			logger.debug("show progress");
			progressBar.setVisibility(View.VISIBLE);
		} else {
			logger.debug("hide progress");
			progressBar.setVisibility(View.GONE);
		}
	}

	@Override
	public void finish() {
		try {
			super.finish();
			overridePendingTransition(0, 0);
		} catch (Exception ignored) {}
	}
}

File: app/src/main/java/ch/threema/app/globalsearch/GlobalSearchActivity.kt
package ch.threema.app.globalsearch

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Menu
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.IdRes
import androidx.appcompat.widget.SearchView
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import ch.threema.android.buildActivityIntent
import ch.threema.app.R
import ch.threema.app.activities.ThreemaToolbarActivity
import ch.threema.app.fragments.composemessage.ComposeMessageFragment.EXTRA_OVERRIDE_BACK_TO_HOME_BEHAVIOR
import ch.threema.app.preference.service.PreferenceService
import ch.threema.app.services.ContactService
import ch.threema.app.services.ConversationCategoryService
import ch.threema.app.services.GroupService
import ch.threema.app.services.MessageService.MessageFilterFlags
import ch.threema.app.services.MessageServiceImpl.FILTER_CHATS
import ch.threema.app.services.MessageServiceImpl.FILTER_GROUPS
import ch.threema.app.services.MessageServiceImpl.FILTER_INCLUDE_ARCHIVED
import ch.threema.app.services.UserService
import ch.threema.app.services.ballot.BallotService
import ch.threema.app.ui.EmptyRecyclerView
import ch.threema.app.ui.EmptyView
import ch.threema.app.ui.InsetSides
import ch.threema.app.ui.SpacingValues
import ch.threema.app.ui.ThreemaSearchView
import ch.threema.app.ui.applyDeviceInsetsAsPadding
import ch.threema.app.utils.ConfigUtils
import ch.threema.app.utils.ContactUtil
import ch.threema.app.utils.GroupUtil
import ch.threema.app.utils.IntentDataUtil
import ch.threema.app.utils.logScreenVisibility
import ch.threema.base.utils.getThreemaLogger
import ch.threema.storage.models.AbstractMessageModel
import ch.threema.storage.models.group.GroupMessageModel
import com.bumptech.glide.Glide
import com.google.android.material.chip.Chip
import com.google.android.material.search.SearchBar
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel

private val logger = getThreemaLogger("GlobalSearchActivity")

class GlobalSearchActivity : ThreemaToolbarActivity(), SearchView.OnQueryTextListener {
    init {
        logScreenVisibility(logger)
    }

    private val contactService: ContactService by inject()
    private val userService: UserService by inject()
    private val groupService: GroupService by inject()
    private val conversationCategoryService: ConversationCategoryService by inject()
    private val ballotService: BallotService by inject()
    private val preferenceService: PreferenceService by inject()
    private val globalSearchViewModel: GlobalSearchViewModel by viewModel()

    private var chatsAdapter: GlobalSearchAdapter? = null
    private var searchView: ThreemaSearchView? = null
    private var searchBar: SearchBar? = null

    @MessageFilterFlags
    private var filterFlags = FILTER_CHATS or FILTER_GROUPS or FILTER_INCLUDE_ARCHIVED
    private var queryText: String? = null
    private val queryHandler = Handler(Looper.getMainLooper())
    private val queryTask = Runnable {
        globalSearchViewModel.onQueryChanged(
            query = queryText,
            filterFlags = filterFlags,
        )
        chatsAdapter?.onQueryChanged(queryText)
    }
    private val showMessageLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { _: ActivityResult ->
            // search results may have changed when returning from ComposeMessageFragment
            globalSearchViewModel.onDataChanged()
        }

    override fun onQueryTextSubmit(query: String?): Boolean {
        return true
    }

    override fun onQueryTextChange(newText: String?): Boolean {
        queryText = newText
        if (chatsAdapter != null) {
            queryHandler.removeCallbacksAndMessages(null)
            if ((queryText?.length ?: 0) >= QUERY_MIN_LENGTH) {
                queryHandler.postDelayed(queryTask, GLOBAL_SEARCH_QUERY_TIMEOUT_MS)
            } else {
                globalSearchViewModel.onQueryChanged(
                    query = null,
                    filterFlags = filterFlags,
                )
                chatsAdapter?.onQueryChanged(null)
            }
        }
        return true
    }

    override fun getLayoutResource(): Int = R.layout.activity_global_search

    override fun handleDeviceInsets() {
        super.handleDeviceInsets()

        findViewById<EmptyRecyclerView>(R.id.recycler_chats).applyDeviceInsetsAsPadding(
            insetSides = InsetSides.lbr(),
        )
    }

    override fun initActivity(savedInstanceState: Bundle?): Boolean {
        if (!super.initActivity(savedInstanceState)) {
            return false
        }

        if (supportActionBar != null) {
            searchBar = toolbar as SearchBar
            searchBar?.let { bar ->
                bar.setNavigationOnClickListener {
                    searchView?.let {
                        if (it.isIconified) {
                            finish()
                        } else {
                            it.isIconified = true
                        }
                    }
                }
                bar.setOnClickListener { searchView?.isIconified = false }
            }
        }

        chatsAdapter = GlobalSearchAdapter(
            this,
            Glide.with(this),
            SNIPPET_THRESHOLD,
            groupService,
            contactService,
            userService,
            ballotService,
            conversationCategoryService,
            preferenceService,
        )
        chatsAdapter?.setOnClickItemListener { messageModel: AbstractMessageModel, _, _ ->
            logger.info("Message clicked")
            showMessage(messageModel)
        }

        setupChip(R.id.chats, FILTER_CHATS)
        setupChip(R.id.groups, FILTER_GROUPS)
        setupChip(R.id.archived, FILTER_INCLUDE_ARCHIVED)

        val recyclerView = this.findViewById<EmptyRecyclerView>(R.id.recycler_chats)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.itemAnimator = DefaultItemAnimator()
        val emptyView = EmptyView(this, ConfigUtils.getActionBarSize(this))
        emptyView.setup(
            R.string.global_search_empty_view_text,
            R.drawable.ic_search_outline,
        )
        (recyclerView.parent.parent as ViewGroup).addView(emptyView)
        recyclerView.emptyView = emptyView
        emptyView.setLoading(true)
        recyclerView.adapter = chatsAdapter

        emptyView.applyDeviceInsetsAsPadding(
            insetSides = InsetSides.lbr(),
            ownPadding = SpacingValues.symmetric(
                vertical = R.dimen.grid_unit_x1,
                horizontal = R.dimen.grid_unit_x3,
            ),
        )

        lifecycleScope.launch {
            globalSearchViewModel.messageModels.collect { messageModels ->
                val messageModelsWithoutPrivateChats = messageModels.filterNot { messageModel ->
                    val uniqueIdString: String =
                        if (messageModel is GroupMessageModel) {
                            GroupUtil.getUniqueIdString(messageModel.groupId.toLong())
                        } else {
                            ContactUtil.getUniqueIdString(messageModel.identity)
                        }
                    conversationCategoryService.isPrivateChat(uniqueIdString)
                }

                chatsAdapter?.setMessageModels(messageModelsWithoutPrivateChats)

                if (messageModelsWithoutPrivateChats.isEmpty()) {
                    if ((queryText?.length ?: 0) >= QUERY_MIN_LENGTH) {
                        emptyView.setup(R.string.search_no_matches)
                    } else {
                        emptyView.setup(R.string.global_search_empty_view_text)
                    }
                }
            }
        }

        lifecycleScope.launch {
            globalSearchViewModel.isLoading.collect { isLoading ->
                emptyView.setLoading(isLoading)
            }
        }

        onQueryTextChange(null)
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.activity_global_search, menu)
        val searchMenuItem = menu.findItem(R.id.menu_action_search)
        searchView = searchMenuItem.actionView as ThreemaSearchView?
        searchView?.let {
            if (ConfigUtils.isLandscape(this)) {
                it.maxWidth = Int.MAX_VALUE
            }

            ConfigUtils.adjustSearchViewPadding(it)
            it.queryHint = getString(R.string.global_search_empty_view_text)
            it.setOnQueryTextListener(this)
            it.setOnSearchClickListener {
                searchBar?.hint = ""
            }
            // Show the hint of the search bar again when the search view is closed
            it.setOnCloseListener {
                searchBar?.setHint(R.string.global_search)
                false
            }
        }
        if (searchView == null) {
            searchMenuItem.isVisible = false
        }
        return true
    }

    private fun setupChip(@IdRes id: Int, flag: Int) {
        // https://github.com/material-components/material-components-android/issues/1419
        val chip = findViewById<Chip>(id)
        chip.isChecked = true
        chip.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            filterFlags = if (isChecked) {
                filterFlags or flag
            } else {
                filterFlags and flag.inv()
            }
            globalSearchViewModel.onQueryChanged(
                query = queryText,
                filterFlags = filterFlags,
            )
        }
    }

    private fun showMessage(messageModel: AbstractMessageModel?) {
        if (messageModel == null) {
            return
        }
        hideKeyboard()
        val intent = IntentDataUtil.getJumpToMessageIntent(this, messageModel)
        intent.putExtra(EXTRA_OVERRIDE_BACK_TO_HOME_BEHAVIOR, true)
        showMessageLauncher.launch(intent)
    }

    private fun hideKeyboard() {
        searchView?.let {
            WindowCompat.getInsetsController(window, it).hide(WindowInsetsCompat.Type.ime())
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        hideKeyboard()
        super.onConfigurationChanged(newConfig)
    }

    companion object {
        private const val QUERY_MIN_LENGTH = 2
        private const val GLOBAL_SEARCH_QUERY_TIMEOUT_MS = 500L
        private const val SNIPPET_THRESHOLD = 17

        fun createIntent(context: Context) = buildActivityIntent<GlobalSearchActivity>(context)
    }
}
